File: design_singly_linked_list/02_pedagogic_oop/python/main.py
from typing import Optional, List

class Node:
    def __init__(self, val: int):
        self.val = val
        self.next: Optional['Node'] = None

class LinkedList:
    """
    02 Pedagogic: Encapsulated Linked List
    """
    def __init__(self):
        self._head: Optional[Node] = None
        self._tail: Optional[Node] = None
        self._size: int = 0

    def insert_head(self, val: int) -> None:
        new_node = Node(val)
        new_node.next = self._head
        self._head = new_node
        if self._size == 0:
            self._tail = new_node
        self._size += 1

    def insert_tail(self, val: int) -> None:
        new_node = Node(val)
        if self._size == 0:
            self._head = new_node
            self._tail = new_node
        else:
            self._tail.next = new_node
            self._tail = new_node
        self._size += 1

    def get_values(self) -> List[int]:
        res = []
        curr = self._head
        while curr:
            res.append(curr.val)
            curr = curr.next
        return res

def main():
    print("=== 02 (Pedagogic OOP) Python ===")
    
    ll = LinkedList()
    ll.insert_head(10)
    ll.insert_tail(20)
    ll.insert_tail(30)
    
    print("List:", ll.get_values())

if __name__ == "__main__":
    main()

File: design_singly_linked_list/02_pedagogic_oop/main.cpp
#include "LinkedList.h"
#include <iostream>

int main() {
  std::cout << "=== 02 (Pedagogic OOP) ===\n";

  teaching::LinkedList list;
  list.insertHead(10);
  list.insertTail(20);
  list.insertTail(30);

  std::cout << "Values: ";
  for (int v : list.getValues())
    std::cout << v << " ";
  std::cout << "\n";

  // Copy Constructor Test
  {
    teaching::LinkedList copy = list;
    copy.insertTail(999);
    std::cout << "Copy should have 999: ";
    for (int v : copy.getValues())
      std::cout << v << " ";
    std::cout << "\n";
  }

  std::cout << "Original should NOT have 999: ";
  for (int v : list.getValues())
    std::cout << v << " ";
  std::cout << "\n";

  return 0;
}
